hail/python/hailtop/batch/job.py

```diff
@@ -600,48 +657,50 @@ def command(self, command: str) -> 'BashJob':
-            if groups['PYTHON_RESULT']:
-                raise BatchException(f"found a reference to a PythonResult object. hint: Use one of the methods `as_str`, `as_json` or `as_repr` on a PythonResult. command: '{command}'")
+        subst_command = self._interpolate_command(command)
+        self._command.append(subst_command)
+        return self
 
-            assert groups['RESOURCE_FILE'] or groups['RESOURCE_GROUP']
-            r_uid = match_obj.group()
-            r = self._batch._resource_map.get(r_uid)
+    async def _compile(self, local_tmpdir, remote_tmpdir):
+        job_command = [cmd.strip() for cmd in self._command]
+        job_command = [f'{{\n{x}\n}}' for x in job_command]
+        job_command = '\n'.join(job_command)
 
-            if r is None:
-                raise BatchException(f"undefined resource '{r_uid}' in command '{command}'.\n"
-                                     f"Hint: resources must be from the same batch as the current job.")
+        if len(job_command.encode()) <= 10 * 1024:
+            self._wrapper_code.append(job_command)
+            return False
 
-            if r._source != self:
-                self._add_inputs(r)
-                if r._source is not None:
-                    if r not in r._source._valid:
-                        name = r._source._resources_inverse[r]
-                        raise BatchException(f"undefined resource '{name}'\n"
-                                             f"Hint: resources must be defined within "
-                                             f"the job methods 'command' or 'declare_resource_group'")
-                    self._dependencies.add(r._source)
-                    r._source._add_internal_outputs(r)
-            else:
-                _add_resource_to_set(self._valid, r)
-            self._mentioned.add(r)
-            return f"${{{r_uid}}}"
+        job_shell = self._shell if self._shell else DEFAULT_SHELL
 
-        regexes = [_resource.ResourceFile._regex_pattern,
-                   _resource.ResourceGroup._regex_pattern,
-                   _resource.PythonResult._regex_pattern,
-                   Job._regex_pattern,
-                   batch.Batch._regex_pattern]
-        subst_command = re.sub('(' + ')|('.join(regexes) + ')',
-                               handler,
-                               command)
-        self._command.append(subst_command)
-        return self
+        pipe = StringIO()
+        pipe.write(f'#! {job_shell}\n')
+        pipe.writelines(job_command)
+        pipe.write('\n')
+        pipe.seek(0)
+
+        code_bytes = pipe.getvalue().encode()
+
+        assert self._job_id is not None
+        job_path = f'{remote_tmpdir}/{self._job_id}'
+        code_path = f'{job_path}/code.sh'
+
+        await self._batch._fs.makedirs(os.path.dirname(code_path), exist_ok=True)
+        await self._batch._fs.write(code_path, code_bytes)
```

Comment:
I have mild performance concerns about copying the whole job command into a StringIO for every job in a large batch. Can we instead do:
```python3
fs = self._batch._fs
async with await retry_transient_errors(fs.create, url) as out:
    await out.write(f'#! {job_shell}\n'.encode())
    await out.write(job_command.encode())
    await out.write(b'\n')
```

Reply:
Oh, I see we need a copy of it for the _user_code anyway. I suppose if we're copying anyway then this is what we must do.